### PR TITLE
Fix retrieve file last commit branchName

### PR DESCRIPTION
**Pull request by lunny**
This will be a part of PR to fix go-gitea/gitea#3181

### CODE REVIEW

**Comment by lafriks** [comment]
LGTM

File: commit_info.go [hunk 1]
```diff
@@ -112,14 +113,15 @@ func initGetCommitInfoState(entries Entries, headCommit *Commit, treePath string
...
+		branchName:    branchName,
 		treePath:      treePath,
 		headCommit:    headCommit,
 	}
 }
 
 // GetCommitsInfo gets information of all commits that are corresponding to these entries
-func (tes Entries) GetCommitsInfo(commit *Commit, treePath string) ([][]interface{}, error) {
-	state := initGetCommitInfoState(tes, commit, treePath)
+func (tes Entries) GetCommitsInfo(commit *Commit, branchName, treePath string) ([][]interface{}, error) {
```
Comment (strk) [inline commit_info.go]: branch name seems redundant here, a commit is enough to identify an entity in a repository. Also changing this signature needlessly changes API of the module

Comment (ethantkoenig) [reply commit_info.go]: Also, sometimes when calling this function there won't be a branch (e.g. viewing the source tree at a particular commit)

Comment (bkcsoft) [reply commit_info.go]: I'd call it `revision` like git does https://git-scm.com/docs/gitrevisions

Comment (lunny) [reply commit_info.go]: @strk maybe right. `HEAD` -> commit.ID.String(). I will fix this.

**Comment by lunny** [comment]
Confirmed release/v1.3 has no this issue.

**Comment by lunny** [comment]
@strk @ethantkoenig also should fix go-gitea/gitea#3202

File: commit_info.go [hunk 2]
```diff
@@ -90,7 +90,9 @@ func targetedSearch(state *getCommitsInfoState, done chan error) {
 			done <- err
 			return
 		}
+		state.lock.Lock()
```
Comment (ethantkoenig) [inline commit_info.go]: Why do we need to lock here? I don't see why it would be a problem if two threads run a read-only git command in parallel.

**Comment by strk** [comment]
Please include a test for the fix

**Comment by lunny** [comment]
@strk will include a test on gitea's integration test.

**Comment by lunny** [comment]
@ethantkoenig I think you are right. Done.

**Comment by appleboy** [comment]
LGTM

**Comment by ethantkoenig** [comment]
@strk I will add tests (which is probably fair, since I introduced the bug)

**Comment by lunny** [comment]
@ethantkoenig maybe add an integration test and compare github's home page tree list and gitea's home page tree list.

**Pull request merged**